include/pybind11/detail/common.h

```diff
@@ -379,12 +387,14 @@ PYBIND11_WARNING_DISABLE_CLANG("-Wgnu-zero-variadic-macro-arguments")
             slots[0] = {Py_mod_exec,                                                              \
                         reinterpret_cast<void *>(&PYBIND11_CONCAT(pybind11_exec_, name))};        \
             slots[1] = {0, nullptr};                                                              \
-            return ::pybind11::module_::initialize_multiphase_module_def(                         \
+            auto o = ::pybind11::module_::initialize_multiphase_module_def(                       \
                 PYBIND11_TOSTRING(name),                                                          \
                 nullptr,                                                                          \
                 &PYBIND11_CONCAT(pybind11_module_def_, name),                                     \
                 slots,                                                                            \
                 ##__VA_ARGS__);                                                                   \
+            PYBIND11_IMMORTALIZE(o);                                                              \
```

Comment:
Do not immortalize objects. The best practice is to use a static pointer instead of a static C++ instance and leak the Python object via `Py_INCREF`.

```cpp
static PyObject *result = new PyObject{[]() -> PyObject* { return init(); }()};
Py_INCREF(result);
return result;
```

Or you can use `py::gil_safe_call_once_and_store`:

```cpp
PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<py::object> storage;
return storage
    .call_once_and_store_result([]() -> py::object {
        auto result = init();
        return result;
    })
    .get_stored();
```

Reply:
Using static C++ instance will call the destructor `~object()`, which calls `Py_DECREF(o)`. However, the Python interpreter has already finalized, and that causes segmentation faults. We should use static pointers over objects.

See also:

- https://github.com/pybind/pybind11/issues/4263#issuecomment-1327510732

Reply:
Is there no way to immortalize an module properly, then? It would be nice to avoid incref/decrefs to the shared refcount across threads; modules would be commonly used across threads.

Reply:
(Seems not, there's some discussion at https://discuss.python.org/t/exposing-public-apis-for-immortal-objects/64264)

Reply:
> Is there no way to immortalize an module properly, then?

We should use pointer instead of instance to avoid `~object()` on program exit. See my comment above https://github.com/pybind/pybind11/pull/5688#discussion_r2105703070.